In [1]:
import csv
import datetime

In [2]:
f = open("guns.csv",'r')
data = list(csv.reader(f))
data[:5]

[['',
  'year',
  'month',
  'intent',
  'police',
  'sex',
  'age',
  'race',
  'hispanic',
  'place',
  'education'],
 ['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4']]

In [3]:
headers = data[0]
headers

['',
 'year',
 'month',
 'intent',
 'police',
 'sex',
 'age',
 'race',
 'hispanic',
 'place',
 'education']

In [4]:
data = data[1:]
data[:5]

[['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'],
 ['5',
  '2012',
  '02',
  'Suicide',
  '0',
  'M',
  '31',
  'White',
  '100',
  'Other specified',
  '2']]

In [5]:
years = []
for row in data:
     years.append(int(row[1]))
years[:5]

[2012, 2012, 2012, 2012, 2012]

In [6]:
year_counts = {}
for year in years:
    if year in year_counts:
        year_counts[year] += 1
    else:
        year_counts[year] = 1

year_counts

{2012: 33563, 2013: 33636, 2014: 33599}

In [7]:
dates = [datetime.datetime(year = int(row[1]),month = int(row[2]), day=1) for row in data]
dates[:5]

[datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 2, 1, 0, 0),
 datetime.datetime(2012, 2, 1, 0, 0)]

In [8]:
date_counts = {}
for date in dates:
    if date.strftime('%B %Y') in date_counts:
        date_counts[date.strftime('%B %Y')] += 1
    else:
        date_counts[date.strftime('%B %Y')] = 1
date_counts

{'April 2012': 2795,
 'April 2013': 2798,
 'April 2014': 2862,
 'August 2012': 2954,
 'August 2013': 2859,
 'August 2014': 2970,
 'December 2012': 2791,
 'December 2013': 2765,
 'December 2014': 2857,
 'February 2012': 2357,
 'February 2013': 2375,
 'February 2014': 2361,
 'January 2012': 2758,
 'January 2013': 2864,
 'January 2014': 2651,
 'July 2012': 3026,
 'July 2013': 3079,
 'July 2014': 2884,
 'June 2012': 2826,
 'June 2013': 2920,
 'June 2014': 2931,
 'March 2012': 2743,
 'March 2013': 2862,
 'March 2014': 2684,
 'May 2012': 2999,
 'May 2013': 2806,
 'May 2014': 2864,
 'November 2012': 2729,
 'November 2013': 2758,
 'November 2014': 2756,
 'October 2012': 2733,
 'October 2013': 2808,
 'October 2014': 2865,
 'September 2012': 2852,
 'September 2013': 2742,
 'September 2014': 2914}

In [9]:
headers

['',
 'year',
 'month',
 'intent',
 'police',
 'sex',
 'age',
 'race',
 'hispanic',
 'place',
 'education']

In [10]:
sex_counts = {}
for row in data:
    if row[5] in sex_counts:
        sex_counts[row[5]] += 1
    else:
        sex_counts[row[5]] = 1
sex_counts

{'F': 14449, 'M': 86349}

In [11]:
def count(idx):
    dic = {}
    for row in data:
        if row[idx] in dic:
            dic[row[idx]] += 1
        else:
            dic[row[idx]] = 1
    return dic
    

In [12]:
race_counts = count(7)
race_counts

{'Asian/Pacific Islander': 1326,
 'Black': 23296,
 'Hispanic': 9022,
 'Native American/Native Alaskan': 917,
 'White': 66237}

## Some details learnt till now

1. No change in Deaths per year
2. Feb showed 2300 deaths than atleast 2700 deaths in other months
3. White people were killed most while Native Americans least
4. Deaths of black were 2nd to most
5. Deaths of Male was to 86,349 than only to 14,449 of females
6. The intent of deaths and education need further examination

In [13]:
census = list(csv.reader(open('census.csv','r')))
census

[['Id',
  'Year',
  'Id',
  'Sex',
  'Id',
  'Hispanic Origin',
  'Id',
  'Id2',
  'Geography',
  'Total',
  'Race Alone - White',
  'Race Alone - Hispanic',
  'Race Alone - Black or African American',
  'Race Alone - American Indian and Alaska Native',
  'Race Alone - Asian',
  'Race Alone - Native Hawaiian and Other Pacific Islander',
  'Two or More Races'],
 ['cen42010',
  'April 1, 2010 Census',
  'totsex',
  'Both Sexes',
  'tothisp',
  'Total',
  '0100000US',
  '',
  'United States',
  '308745538',
  '197318956',
  '44618105',
  '40250635',
  '3739506',
  '15159516',
  '674625',
  '6984195']]

In [14]:
race_census = {}
for key in race_counts:
    race_census[key] = 0
race_census

{'Asian/Pacific Islander': 0,
 'Black': 0,
 'Hispanic': 0,
 'Native American/Native Alaskan': 0,
 'White': 0}

In [15]:
for key,races in enumerate(census[0]):
    print(key,races)

0 Id
1 Year
2 Id
3 Sex
4 Id
5 Hispanic Origin
6 Id
7 Id2
8 Geography
9 Total
10 Race Alone - White
11 Race Alone - Hispanic
12 Race Alone - Black or African American
13 Race Alone - American Indian and Alaska Native
14 Race Alone - Asian
15 Race Alone - Native Hawaiian and Other Pacific Islander
16 Two or More Races


In [16]:
race_census['Asian/Pacific Islander'] = int(census[1][14]) + int(census[1][15])
race_census['Black'] = int(census[1][12])
race_census['Hispanic'] = int(census[1][11])
race_census['Native American/Native Alaskan'] = int(census[1][13])
race_census['White'] = int(census[1][10])
race_census

{'Asian/Pacific Islander': 15834141,
 'Black': 40250635,
 'Hispanic': 44618105,
 'Native American/Native Alaskan': 3739506,
 'White': 197318956}

In [17]:
race_per_hundredk = {}
for key,counts in race_counts.items():
    race_per_hundredk[key] = (counts/race_census[key]) * 100000
race_per_hundredk

{'Asian/Pacific Islander': 8.374309664161762,
 'Black': 57.8773477735196,
 'Hispanic': 20.220491210910907,
 'Native American/Native Alaskan': 24.521955573811088,
 'White': 33.56849303419181}

In [18]:
intents = [row[3] for row in data]
intents[:10]

['Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Undetermined',
 'Suicide',
 'Accidental',
 'Suicide']

In [19]:
races = [row[7] for row in data]


In [20]:
homicide_race_counts = {}
for idx,intent in enumerate(intents):
    if intent == 'Homicide':
        if races[idx] in homicide_race_counts:
            homicide_race_counts[races[idx]] += 1
        else:
            homicide_race_counts[races[idx]] = 1

homicide_race_counts
        
        

{'Asian/Pacific Islander': 559,
 'Black': 19510,
 'Hispanic': 5634,
 'Native American/Native Alaskan': 326,
 'White': 9147}

In [21]:
for key,count in homicide_race_counts.items():
    homicide_race_counts[key] = (count/race_census[key])*100000
homicide_race_counts

{'Asian/Pacific Islander': 3.530346230970155,
 'Black': 48.471284987180944,
 'Hispanic': 12.627161104219914,
 'Native American/Native Alaskan': 8.717729026240365,
 'White': 4.6356417981453335}